stl/inc/algorithm

```diff
@@ -961,6 +941,322 @@ namespace ranges {
+                ranges::equal_to>
+        _NODISCARD constexpr bool operator()(
+            _Rng1&& _Range1, _Rng2&& _Range2, _Pr _Pred = {}, _Pj1 _Proj1 = {}, _Pj2 _Proj2 = {}) const {
+            if constexpr (sized_range<_Rng1> && sized_range<_Rng2>) {
+                const auto _Count = _RANGES distance(_Range1);
+                if (_RANGES distance(_Range2) != _Count) {
+                    return false;
+                }
+
+                return _Sized_ranges(_Get_unwrapped(_RANGES begin(_Range1)), _Get_unwrapped(_RANGES end(_Range1)),
+                    _Get_unwrapped(_RANGES begin(_Range2)), _Get_unwrapped(_RANGES end(_Range2)), _Count,
+                    _Pass_fn(_Pred), _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+            } else {
+                return _Unsized_ranges(_Get_unwrapped(_RANGES begin(_Range1)), _Get_unwrapped(_RANGES end(_Range1)),
+                    _Get_unwrapped(_RANGES begin(_Range2)), _Get_unwrapped(_RANGES end(_Range2)), _Pass_fn(_Pred),
+                    _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+            }
+        }
+
+    private:
+        template <class _It1, class _Se1, class _It2, class _Se2, class _Pr, class _Pj1, class _Pj2>
+        _NODISCARD static constexpr bool _Sized_ranges(_It1 _First1, _Se1 _End1, _It2 _First2, _Se2 _End2,
+            iter_difference_t<_It1> _Count, _Pr _Pred, _Pj1 _Proj1, _Pj2 _Proj2) {
+            _STL_INTERNAL_CHECK(_RANGES distance(_First1, _End1) == _Count);
+            _STL_INTERNAL_CHECK(_RANGES distance(_First2, _End2) == _Count);
+
+            for (;; ++_First1, (void) ++_First2, --_Count) { // Trim matching prefixes
+                if (_Count == 0) { // Everything matched
+                    return true;
+                }
+
+                if (!_STD invoke(_Pred, _STD invoke(_Proj1, *_First1), _STD invoke(_Proj2, *_First2))) { // mismatch
+                    if (_Count == 1) { // Single non-matching characters remain; not a permutation
+                        return false;
+                    }
+
+                    break;
+                }
+            }
+            // If we get here, _Count > 1 and initial elements do not match.
+
+            if constexpr (bidirectional_iterator<_It1> && bidirectional_iterator<_It2>) {
+                // Determine final iterator values
+                auto _Last1 = _Find_last_iterator(_First1, _End1, _Count);
+                auto _Last2 = _Find_last_iterator(_First2, _End2, _Count);
+
+                for (;;) { // Trim matching suffixes
+                    --_Last1;
+                    --_Last2;
+                    if (!_STD invoke(_Pred, _STD invoke(_Proj1, *_Last1), _STD invoke(_Proj2, *_Last2))) { // mismatch
+                        ++_Last1;
+                        ++_Last2;
+                        break;
+                    }
+
+                    if (--_Count == 1) {
+                        return false; // initial elements did not match when trimming prefixes

```

Comment:
This comment, which occurs below, is helpful but slightly confusing. Specifically, "when trimming prefixes" makes it sound like "initial elements did not match" at the beginning of that process (especially because of "initial"). I think this would be clearer if it said "after trimming prefixes".

Reply:
I think "initial elements still do not match" more directly makes the point that we proved this fact earlier.